Aerodynamic forces correction enhancements

ngoiz opened the pull request:
This PR tranfsorms the aerodynamic forces correction functions to generators to be used in the coupled solvers. This is for cases where airfoil polars are defined (see updated Case Files documentation). 

In this PR we include:
 * Correct forces: transform functions to generators
 * Bug fixes in polar corrections and full documentation
 * Update to the `aeroforcescalculator` post-processor to output corrected forces and to output aerodynamic forces and moments. The latter are now written on a separate file `moments_`
 * Changes to the aero data structure such that the aero forces expressed in the beam degrees of freedom are included in the aero time step

sduess commented on sharpy/aero/utils/mapping.py:
```diff
@@ -75,3 +75,55 @@ def aero2struct_force_mapping(aero_forces,
+    Args:
+        aero_forces (list): Aerodynamic forces from the UVLM in inertial frame of reference
+        struct2aero_mapping (dict): Structural to aerodynamic node mapping
+        zeta (list): Aerodynamic grid coordinates
+        pos_def (np.ndarray): Vector of structural node displacements
+        psi_def (np.ndarray): Vector of structural node rotations (CRVs)
+        conn (np.ndarray): Connectivities matrix
+        cag (np.ndarray): Transformation matrix between inertial and body-attached reference ``A``
+        aero_dict (dict): Dictionary containing the grid's information.
+
+    Returns:
+        np.array: Vector ``6`` containing the total aerodynamic forces and moments at the A frame, expressed
+          in ``A``.
+    """
+    n_node, _ = pos_def.shape
+    n_elem, _, _ = psi_def.shape
+    total_aero_forces = np.zeros(6)
+
+    nodes = []
+
+    for i_elem in range(n_elem):
+        for i_local_node in range(3):
+
+            i_global_node = conn[i_elem, i_local_node]
+            if i_global_node in nodes:
+                continue
+
+            nodes.append(i_global_node)
+            for mapping in struct2aero_mapping[i_global_node]:
+                i_surf = mapping['i_surf']
+                i_n = mapping['i_n']
+                _, n_m, _ = aero_forces[i_surf].shape
+
+                for i_m in range(n_m):
+                    zeta_g = zeta[i_surf][:, i_m, i_n]
+                    total_aero_forces[0:3] += cag.dot(aero_forces[i_surf][0:3, i_m, i_n])
+                    total_aero_forces[3:6] += cag.dot(aero_forces[i_surf][3:6, i_m, i_n])
+                    total_aero_forces[3:6] += cag.dot(algebra.cross3(zeta_g, aero_forces[i_surf][0:3, i_m, i_n]))
+
+    return total_aero_forces
```
```suggestion

```
The function is not used anywhere. You probably have just forgotten to remove it.

ngoiz replied:
Indeed! Thanks for spotting it!

sduess commented on sharpy/utils/datastructures.py:
```diff
@@ -165,13 +165,15 @@ def __init__(self, dimensions, dimensions_star):
         self.inertial_unsteady_forces = np.zeros((self.n_surf, 6))
         self.body_unsteady_forces = np.zeros((self.n_surf, 6))
+        self.total_inertial_forces = np.zeros((6,))  # G Frame
+        self.total_body_forces = np.zeros((6,))  # A Frame
```
These forces are not used anywhere as well are others. You committed the necessary changes in commit 8d1021e30e195a79795c6ed8357c1ac286367e0b

sduess reviewed (changes requested):
Good job Norberto! I just added some minor suggestions which are all related to "delete stuff we don't use in our code".

ArturoMS13 reviewed (changes requested):
Thank you very much for the contribution. I add some minor comments open to discussion.

ngoiz commented on docs/source/content/casefiles.rst:
```diff
@@ -315,6 +315,14 @@ Item by item:
+
+    This is an optional group to add if correcting the aerodynamic forces using airfoil polars is desired. A polar
+    should be included for each airfoil defined. Each entry consists of a 4-column table. The first column corresponds
+    to the angle of attack (in radians) and then the ``C_L``, ``C_D`` and ``C_M``. If different airfoils are used a
```
True! 😅

ngoiz commented:
Thanks both for your comments. I think I have addressed all your points:
* Horrible `aerotstep.aero_forces_beam_dof` is gone and instead it goes into `structtstep.postproc_nodes['aero_steady_forces]` (or unsteady) which is then plotted in Paraview
* Opted to retain `aero_tstep.body_steady_forces` since those are a `6x` vector and for each surface, which is useful in some cases and it doesn't go in nicely onto the `postproc_node` strategy as above. These are defined in the `AeroTimestepInfo` class.
* Adapted the `AeroForcesCalculator` postproc to run online and offline
* Moved the test to the polars tests `/tests/uvlm/static/polars`

I think this covers my changes. @ArturoMS13 once you are happy with the stability axes and your twisting wings I think this should be then good to go

ArturoMS13 commented:
Thank you for the contribution
I have double checked this implementation for wind turbines and it works except from the root regions. I will fix the root soon, but it is not a big deal so this PR can be merged.

ArturoMS13 commented on sharpy/generators/polaraeroforces.py:
```diff
@@ -0,0 +1,286 @@
+          (the remaining force).
+
+    If ``cd_from_cl == 'on'``.
+        2. The viscous drag and pitching moment is found at the computed lift coefficient and the forces and
+           moments updated
```
```suggestion
        2. The viscous drag and pitching moment are found at the computed lift coefficient. Then forces and
           moments are updated
```

ArturoMS13 commented on sharpy/generators/polaraeroforces.py:
```diff
@@ -0,0 +1,286 @@
+    Else, the angle of attack is computed:
+
+            2. The angle of attack is computed based on that lift force and the angle of zero lift computed from the
+               airfoil polar and taking the potential flow lift curve slope of :math:`2 \pi`
```
```suggestion
               airfoil polar and assuming the potential flow lift curve slope of :math:`2 \pi`
```

ArturoMS13 commented on sharpy/generators/polaraeroforces.py:
```diff
@@ -0,0 +1,286 @@
+        4. If ``correct_lift == 'on'``, the lift coefficient is also corrected with the polar data. Else only the
+           UVLM results are used.
+
+    The pitching moment is added in a similar manner to the viscous drag. However, if ``moment_from_polar == 'on'``
```
```suggestion
    The pitching moment is added in a similar manner as the viscous drag. However, if ``moment_from_polar == 'on'``
```

ArturoMS13 commented on sharpy/generators/polaraeroforces.py:
```diff
@@ -0,0 +1,286 @@
+    settings_types['moment_from_polar'] = 'bool'
+    settings_default['moment_from_polar'] = False
+    settings_description['moment_from_polar'] = 'If ``correct_lift`` is selected, it will compute the pitching moment ' \
+                                                'simply from poalr derived data, i.e. the polars Cm and the moments' \
```
```suggestion
                                                'simply from polar derived data, i.e. the polars Cm and the moments' \
```

ArturoMS13 commented on sharpy/generators/polaraeroforces.py:
```diff
@@ -0,0 +1,286 @@
+
+        3. The drag force is computed based on the angle of attack and the polars provided by the user
+
+        4. If ``correct_lift == 'on'``, the lift coefficient is also corrected with the polar data. Else only the
```
```suggestion
        4. If ``correct_lift == 'on'``, the lift coefficient is also corrected with the polar data. Else, only the
```

ArturoMS13 reviewed (approved):
Just some extra minor changes. Thank you for the contribution
